core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -159,13 +166,14 @@ static void prepareHeaders(Metadata headers, DecompressorRegistry decompressorRe
   @Override
   public void start(final Listener<RespT> observer, Metadata headers) {
     checkState(stream == null, "Already started");
-    checkNotNull(observer, "observer");
-    checkNotNull(headers, "headers");
+    this.observer = checkNotNull(observer, "observer");
+    this.headers = checkNotNull(headers, "headers");
 
     if (context.isCancelled()) {
       // Context is already cancelled so no need to create a real stream, just notify the observer
       // of cancellation via callback on the executor
       stream = NoopClientStream.INSTANCE;
+      streamStarted = true;
```

Comment:
Can `streamStarted` be made less error-prone? If setting it is necessary here (and I can't quite tell if it is) then it would also seem necessary in the `compressor == null` case below and when `deadlineExceeded == true`.

I think I would prefer most cases using `streamStarted` to use `useGetMethod` instead, as it doesn't change state as randomly. Just `startStream()` and `cancelled()` would use `streamStarted` and just for internal validation.